### PR TITLE
Refactor/extract page properties

### PR DESCRIPTION
**Pull request by RobjS**
This PR refactors the `SendHeaders` class to extract the page data out to a separate `Page` model that holds all the information about what kind of page is currently being viewed. This is hopefully the first step in a large refactor to make the plugin's code more modular and developer-friendly.

Essentially, the refactor takes all the logic that was previously put into the `SendHeaders::pageProperties` property and moves it to the `Page` class. This is demonstrated in commit `f42b7b9`, which implements the new class, but does not amend the tests to mock the calls to the new class's methods. Instead, the existing test mocks are retained (which mock the logic that now happens within `Page`'s methods). The fact that the tests all continued to pass demonstrates that no changes to the overall behaviour have occurred. The tests were then refactored fully in the following commit, so that the tests are only examining the single `SendHeaders` class under test.

I have fixed a couple of minor bugs that emerged during the refactor, where we were expecting the wrong kind of return type if `get_post_type()` or `get_post_taxonomies()` fail, but these are edge cases that should not affect standard behaviour of the plugin.

## How to test

1. Run `composer install && vendor/bin/php-cs-fixer fix --dry-run & vendor/bin/psalm && vendor/bin/kahlan spec` to run the tests
2. Install and activate this branch of the plugin on a test site and do some general testing, particularly to ensure that `no-cache` headers are always served if you're e.g. logged in, or viewing a preview

### CODE REVIEW

File: spec/send_headers.spec.php [hunk 1]
```diff
@@ -6,6 +6,8 @@
 	beforeEach(function () {
 		$this->sendHeaders = new \CacheControl\SendHeaders();
 
+		global $post;
+		$post = (object) [];
```
Comment (snim2) [inline spec/send_headers.spec.php:13]: Should this use `StdClass`? (I notice sometimes we do, sometimes we don't...)

File: src/SendHeaders.php [hunk 2]
```diff
@@ -9,12 +9,17 @@ class SendHeaders implements \Dxw\Iguana\Registerable
 	protected bool $developerMode = false;
 	protected bool $overriddenByTaxonomy = false;
 	protected string $currentConfig = 'default';
-	protected array $pageProperties = [];
+	protected object $page;
```
Comment (snim2) [inline src/SendHeaders.php:12]: Why not `protected Page $page`? Are we not using the new class yet?

**Review by snim2 — APPROVED** [review]
This looks sensible and Kahlan runs green for me. I've done a little manual testing on a local env (but not much) and it seems OK. 

Note that to run the tests through `phpdbg` you now need to manually adjust the memory limit as this branch is much more resource hungry than `main`.

Without adjusting the max memory:

```shell
❯ phpdbg  -qrr ./vendor/bin/kahlan --coverage=4 --no-colors
            _     _
  /\ /\__ _| |__ | | __ _ _ __
 / //_/ _` | '_ \| |/ _` | '_ \
/ __ \ (_| | | | | | (_| | | | |
\/  \/\__,_|_| |_|_|\__,_|_| |_|

The PHP Test Framework for Freedom, Truth and Justice.

src directory  : /Users/sarah.mount/code/github.com/dxw/saluki-test-site/wp-content/plugins/dxw-cache-control/src
spec directory : /Users/sarah.mount/code/github.com/dxw/saluki-test-site/wp-content/plugins/dxw-cache-control/spec

................................................................. 65 / 90 ( 72%)
...................
[PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 65536 bytes) in .../dxw-cache-control/vendor/kahlan/kahlan/src/Jit/TokenStream.php on line 186]
```


and with:

```shell
❯ phpdbg -d memory_limit=-1 -qrr ./vendor/bin/kahlan --coverage=4 --no-colors
            _     _
  /\ /\__ _| |__ | | __ _ _ __
 / //_/ _` | '_ \| |/ _` | '_ \
/ __ \ (_| | | | | | (_| | | | |
\/  \/\__,_|_| |_|_|\__,_|_| |_|

The PHP Test Framework for Freedom, Truth and Justice.

src directory  : .../dxw-cache-control/src
spec directory : .../dxw-cache-control/spec

................................................................. 65 / 90 ( 72%)
.........................                                         90 / 90 (100%)



Expectations   : 231 Executed
Specifications : 0 Pending, 0 Excluded, 0 Skipped

Passed 90 of 90 PASS in 0.534 seconds (using 154MB)

Coverage Summary
----------------
                                                    Lines             %

 \                                                135 / 149      90.60%
└── CacheControl\                                 135 / 144      93.75%
   ├── Options                                     44 /  44     100.00%
   │  ├── Options::addActionLinks()                 3 /   3     100.00%
   │  ├── Options::addOptions()                     9 /   9     100.00%
   │  ├── Options::addOptionsPage()                 1 /   1     100.00%
   │  ├── Options::getPostTypesConfig()            12 /  12     100.00%
   │  ├── Options::getTaxonomiesConfig()           10 /  10     100.00%
   │  ├── Options::getTemplatesConfig()             6 /   6     100.00%
   │  └── Options::register()                       3 /   3     100.00%
   ├── Page                                        13 /  14      92.86%
   │  ├── Page::isAdmin()                           1 /   1     100.00%
   │  ├── Page::isArchivePage()                     1 /   1     100.00%
   │  ├── Page::isFrontPage()                       1 /   1     100.00%
   │  ├── Page::isHomePage()                        1 /   1     100.00%
   │  ├── Page::isLoggedInUser()                    1 /   1     100.00%
   │  ├── Page::isPreviewPage()                     1 /   1     100.00%
   │  ├── Page::postId()                            3 /   3     100.00%
   │  ├── Page::postType()                          1 /   1     100.00%
   │  ├── Page::requiresPassword()                  1 /   2      50.00%
   │  ├── Page::taxonomies()                        1 /   1     100.00%
   │  └── Page::templateName()                      1 /   1     100.00%
   └── SendHeaders                                 78 /  86      90.70%
      ├── SendHeaders::__construct()                1 /   1     100.00%
      ├── SendHeaders::getContext()                 2 /   2     100.00%
      ├── SendHeaders::getPageConfiguration()      46 /  54      85.19%
      ├── SendHeaders::outputDeveloperMeta()       11 /  11     100.00%
      ├── SendHeaders::register()                   2 /   2     100.00%
      └── SendHeaders::setCacheHeader()            16 /  16     100.00%

Total: 90.60% (135/149)

Coverage collected in 0.017 seconds
```

Note that on `main` without adjusting the memory limit you get very slightly more code coverage:

```shell
Passed 73 of 73 PASS in 0.070 seconds (using 25MB)

Coverage Summary
----------------
                                                    Lines             %

 \                                                126 / 139      90.65%
└── CacheControl\                                 126 / 135      93.33%
   ├── Options                                     44 /  44     100.00%
   │  ├── Options::addActionLinks()                 3 /   3     100.00%
   │  ├── Options::addOptions()                     9 /   9     100.00%
   │  ├── Options::addOptionsPage()                 1 /   1     100.00%
   │  ├── Options::getPostTypesConfig()            12 /  12     100.00%
   │  ├── Options::getTaxonomiesConfig()           10 /  10     100.00%
   │  ├── Options::getTemplatesConfig()             6 /   6     100.00%
   │  └── Options::register()                       3 /   3     100.00%
   └── SendHeaders                                 82 /  91      90.11%
      ├── SendHeaders::getContext()                 2 /   2     100.00%
      ├── SendHeaders::getPageConfiguration()      46 /  54      85.19%
      ├── SendHeaders::getPageProperties()         12 /  12     100.00%
      ├── SendHeaders::getPostId()                  3 /   3     100.00%
      ├── SendHeaders::hasPassword()                1 /   2      50.00%
      ├── SendHeaders::register()                   2 /   2     100.00%
      └── SendHeaders::setCacheHeader()            16 /  16     100.00%

Total: 90.65% (126/139)

Coverage collected in 0.015 seconds
```

**Comment by RobjS** [comment]
> 
> Note that on `main` without adjusting the memory limit you get very slightly more code coverage:
> 
> ```shell
> Passed 73 of 73 PASS in 0.070 seconds (using 25MB)
> 
> Coverage Summary
> ----------------
>                                                     Lines             %
> 
>  \                                                126 / 139      90.65%
> └── CacheControl\                                 126 / 135      93.33%
>    ├── Options                                     44 /  44     100.00%
>    │  ├── Options::addActionLinks()                 3 /   3     100.00%
>    │  ├── Options::addOptions()                     9 /   9     100.00%
>    │  ├── Options::addOptionsPage()                 1 /   1     100.00%
>    │  ├── Options::getPostTypesConfig()            12 /  12     100.00%
>    │  ├── Options::getTaxonomiesConfig()           10 /  10     100.00%
>    │  ├── Options::getTemplatesConfig()             6 /   6     100.00%
>    │  └── Options::register()                       3 /   3     100.00%
>    └── SendHeaders                                 82 /  91      90.11%
>       ├── SendHeaders::getContext()                 2 /   2     100.00%
>       ├── SendHeaders::getPageConfiguration()      46 /  54      85.19%
>       ├── SendHeaders::getPageProperties()         12 /  12     100.00%
>       ├── SendHeaders::getPostId()                  3 /   3     100.00%
>       ├── SendHeaders::hasPassword()                1 /   2      50.00%
>       ├── SendHeaders::register()                   2 /   2     100.00%
>       └── SendHeaders::setCacheHeader()            16 /  16     100.00%
> 
> Total: 90.65% (126/139)
> 
> Coverage collected in 0.015 seconds
> ```
I *think* this is just a result of the increase in the number of lines of code, rather than an actual reduction in test coverage? Though I can't work out why it thinks `Page::requiresPassword()` only has 50% coverage...
